torch_frame/nn/encoder/stype_encoder.py

```diff
@@ -685,3 +725,71 @@ def encode_forward(self, feat: TensorData) -> Tensor:
         # -> [batch_size, num_cols, out_channels]
         x = x_lin + self.bias
         return x
+
+
+class TimestampEncoder(StypeEncoder):
+    r"""TimestampEncoder for timestamp stype. The encoder applies
+    different frequencies to the timestamps to calculate the
+    positional encoding and then adds the positional encoding to
+    the original :obj:`Tensor`. Original encoder is introduced in
+    `"Attention Is All You Need"
+    <https://arxiv.org/abs/1706.03762>`_ paper.
+
+    Args:
+        num_freqs: Number of times to apply the different sine and
+        cosine frequencies.
+    """
+    supported_stypes = {stype.timestamp}
+
+    def __init__(
+        self,
+        out_channels: Optional[int] = None,
+        stats_list: Optional[List[Dict[StatType, Any]]] = None,
+        stype: Optional[stype] = None,
+        post_module: Optional[Module] = None,
+        na_strategy: Optional[NAStrategy] = None,
+        num_freqs: int = 8,
+    ):
+        self.num_freqs = num_freqs
+        super().__init__(out_channels, stats_list, stype, post_module,
+                         na_strategy)
+
+    def init_modules(self):
+        super().init_modules()
+        min_year = torch.tensor([
+            self.stats_list[i][StatType.YEAR_RANGE][0]
+            for i in range(len(self.stats_list))
+        ])
+        self.register_buffer("min_year", min_year)
+        div_term = torch.exp(
+            torch.arange(0, self.num_freqs, 2).float() *
+            (-np.log(10000.0) / self.num_freqs))
+        self.register_buffer("div_term", div_term)
+        position = torch.arange(0, stype.timestamp.timestamp_num_cols,
+                                dtype=torch.float)
+        self.register_buffer("position", position)
+        self.linear = Linear(
+            stype.timestamp.timestamp_num_cols * self.num_freqs,
+            self.out_channels)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        super().reset_parameters()
+
+    def encode_forward(self, feat: Tensor) -> Tensor:
+        feat[:, :, TIME_TO_INDEX['YEAR']] = feat[:, :, 0] - self.min_year
+        feat[:, :, TIME_TO_INDEX['MONTH']] /= NUM_MONTHS_PER_YEAR
+        feat[:, :, TIME_TO_INDEX['DAY']] /= NUM_DAYS_PER_MONTH
+        feat[:, :, TIME_TO_INDEX['DAYOFWEEK']] /= NUM_DAYS_PER_WEEK
+        feat[:, :, TIME_TO_INDEX['HOUR']] /= NUM_HOURS_PER_DAY
+        feat[:, :, TIME_TO_INDEX['MINUTE']] /= NUM_MINUTES_PER_HOUR
+        feat[:, :, TIME_TO_INDEX['SECOND']] /= NUM_SECONDS_PER_MINUTE
```

Comment:
Can we avoid the in-place modification here? I think it is not expected that input gets modified in-place.

Reply:
I don't think this is in-place modification. The tensor passed into this function will be cloned first. https://github.com/pyg-team/pytorch-frame/blob/master/torch_frame/nn/encoder/stype_encoder.py#L93

Reply:
Oh, any reason we do a copy there?

Reply:
Because we don't want to modify the original data via NaN imputation.

Reply:
But isn't the original data already copied before the NaN imputation?

Reply:
I don't think so. Where did we copy?

Reply:
[here](https://github.com/pyg-team/pytorch-frame/blob/master/torch_frame/nn/encoder/stype_encoder.py#L115)

Reply:
Yes, I am aware of that. So the in-place imputation is safe, right?